drivers/i2c/i2c_emul_ctlr.c

```diff
@@ -0,0 +1,144 @@
+/*
+ * This driver creates fake I2C buses which can contain emulated devices,
+ * implemented by a separate emulation driver. The API between this driver and
+ * its emulators is defined by struct i2c_emul_driver_api.
+ *
+ * Copyright 2020 Google LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT zephyr_i2c_emul_controller
+
+#define LOG_LEVEL CONFIG_I2C_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(i2c_emul_ctlr);
+
+#include <device.h>
+#include <drivers/i2c.h>
+#include <drivers/i2c_emul.h>
+
+/** Working data for the device */
+struct i2c_emul_data {
+	uint32_t i2c_cfg;
+};
+
+/** Read-only configuration data for the device */
+struct i2c_emul_cfg {
+	uint32_t base;
+};
+
+static sys_slist_t i2c_emul_list;
+
+static struct i2c_emul_registration *i2c_emul_find(int addr)
+{
+	sys_snode_t *node;
+
+	SYS_SLIST_FOR_EACH_NODE(&i2c_emul_list, node) {
+		struct i2c_emul_registration *emul = NULL;
+
+		emul = CONTAINER_OF(node, struct i2c_emul_registration, node);
+		if (emul->addr == addr) {
+			return emul;
+		}
+	}
+
+	return NULL;
+}
+
+static int i2c_emul_configure(struct device *dev, uint32_t dev_config)
+{
+	struct i2c_emul_data *data = dev->driver_data;
+	sys_snode_t *node;
+
+	/*
+	 * Tell all the emulators, since they have no way of looking up this
+	 * information otherwise
```

Comment:
Sure they do.  They know their parent controller, and can get its `struct device*` pointer, and can invoke emulator-specific operations on it.